Core/FunctionTree/Parameter.hpp

```diff
@@ -101,30 +101,14 @@ class Parameter {
   // Observer Pattern Functions
 
   /// Attaches a new TreeNode as Observer
-  void Attach(std::shared_ptr<ParObserver> newObserver) {
+  void Attach(std::weak_ptr<ParObserver> newObserver) {
```

Comment:
So if I understand correctly, the key point is that deletion is not anymore done by all `shared_ptr` running out of scope?

Reply:
Hmm I can't follow your question. I try to explain the core idea shortly.
Before there where `shared_ptr` of objects that had `shared_ptr` members to each other, creating a cyclic reference. That prevents running out of scope to clean up automatically.
The solution is to use a `weak_ptr` for one of the links, since `weak_ptr` avoid increasing the `shared_ptr` count of the object. Then the automatic cleanup works.
Because all of the `weak_ptr` are only used if the objects they are pointing to actually exist (by construction), its safe to use this.
I should verify that the parameters are actually also all deleted, to ensure that the Observers are also guaranteed to exist.

Reply:
Okay thanks, yes that answers my question.

Reply:
Ok so I checked and also this part is working fine. Because when we create an Intensity, all data Parameters and fit Parameters are created completely fresh and are only stored within the Intensity. So when the Intensity object runs out of scope, it is guaranteed by design to clean up nicely without any issues and no old Parameter are ever reused.

Issues could for example occur if someone would change the code that the `PartProps` would return a `shared_ptr` of the Mass of the Particle and if that would be inserted in the FunctionTree on every new construction.

Reply:
The new code also adds a customized destructor for the TreeNode, which unregisters itself from all objects with weak ptrs to it.